### PR TITLE
Migrate Efficientnetv1 to Backbone 

### LINKED ISSUE #1633: Migrate EfficientNetV1 to the new `Backbone` API
To align KerasCV with the KerasNLP API, we're migrating our backbone models to subclass keras_cv.models.Backbone.
This allows us to have a more structured, modular API for high-level task models, including the new ImageClassifier model (see https://github.com/keras-team/keras-cv/pull/1458).

A great step-by-step guide to this migration can be found in the description of https://github.com/keras-team/keras-cv/pull/1508, which performed this migration for ResNetV1.

If you'd like to take this issue up, please leave a comment and the KerasCV team can assign the issue to you. When opening a PR, please tag this issue as fixed by your PR, and tag our tracking bug https://github.com/keras-team/keras-cv/issues/1627 for bookkeeping.

### PR DESCRIPTION
**Pull request by cosmo3769**
# What does this PR do?

Fixes #1633 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? #1631 
- [ ] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

@LukeWood @ianstenbit

### CODE REVIEW

**Comment by cosmo3769** [comment]
Waiting for this #1723 to complete. Will structure it in the same way.

**Comment by cosmo3769** [comment]
Starting to work on this since #1723 is complete.

**Comment by cosmo3769** [comment]
I have made all the required changes to migrate efficientnetv1 to backbone (It is referenced from #1723). Here is the [colab link](https://colab.research.google.com/drive/1vOfQT8ZvTpZd3W0Cv9Fxbs1C1wcV9eoE?usp=sharing) to verify all the tests. It is ready for review. Please review. Thank you. @jbischof @ianstenbit @LukeWood

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+from keras_cv.models.backbones.backbone import Backbone
+from keras_cv.models.backbones.efficientnet_v1.efficientnet_v1_backbone_presets import (  # noqa: E501
+    backbone_presets,
+)
+from keras_cv.utils.python_utils import classproperty
+
+
+@keras.utils.register_keras_serializable(package="keras_cv.models")
+class EfficientNetV1Backbone(Backbone):
+    """This class represents a Keras EfficientNet architecture.
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:40]: Empty line after top level description

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+        kernel_size=1,
+        strides=1,
+        padding="same",
+        use_bias=False,
+        kernel_initializer=conv_kernel_initializer(),
+        name=name + "_output_phase",
+    )(x)
+    x = layers.BatchNormalization(
+        axis=3,
+        name=name + "outout_phase_bn",
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:471]: s/outout/output

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+        depth_coefficient: float, scaling coefficient for network depth.
+        dropout_rate: float, dropout rate before final classifier layer.
+        drop_connect_rate: float, dropout rate at skip connections.
+        depth_divisor: integer, a unit of network width.
+        activation: activation function.
+        input_shape: optional shape tuple, it should have exactly 3 input
+            channels.
+        input_tensor: optional Keras tensor (i.e. output of `layers.Input()`) to
+            use as image input for the model.
+        id_skip: boolean.
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:54]: This arg needs way better documentation. Since it is always `True` in our presets I would strongly consider dropping this entirely.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+class EfficientNetV1Backbone(Backbone):
+    """This class represents a Keras EfficientNet architecture.
+    Args:
+        include_rescaling: bool, whether to rescale the inputs. If set to
+            True, inputs will be passed through a `Rescaling(1/255.0)` layer.
+        width_coefficient: float, scaling coefficient for network width.
+        depth_coefficient: float, scaling coefficient for network depth.
+        dropout_rate: float, dropout rate before final classifier layer.
+        drop_connect_rate: float, dropout rate at skip connections.
+        depth_divisor: integer, a unit of network width.
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:48]: This arg needs better documentation

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+        x = layers.BatchNormalization(
+            axis=3,
+            name="stem_bn",
+        )(x)
+        x = layers.Activation(activation, name="stem_activation")(x)
+
+        # Build blocks
+        block_id = 0
+        blocks = float(
+            sum(num_repeats for num_repeats in stackwise_num_repeats)
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:162]: Why not

```python
sum(stackwise_num_repeats)
```

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+        self.include_rescaling = include_rescaling
+        self.width_coefficient = width_coefficient
+        self.depth_coefficient = depth_coefficient
+        self.dropout_rate = dropout_rate
+        self.drop_connect_rate = drop_connect_rate
+        self.depth_divisor = depth_divisor
+        self.activation = activation
+        self.input_tensor = input_tensor
+        self.pyramid_level_inputs = {
+            i + 1: name for i, name in enumerate(pyramid_level_inputs)
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:253]: We're now using strings for keys. See EfficientNetV1 ([link](https://github.com/keras-team/keras-cv/blob/deefe948bb2f9d72695b404b9ee1caa35ee16482/keras_cv/models/backbones/efficientnet_v2/efficientnet_v2_backbone.py#L260))

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+
+
+def conv_kernel_initializer(scale=2.0):
+    return keras.initializers.VarianceScaling(
+        scale=scale, mode="fan_out", distribution="truncated_normal"
+    )
+
+
+def correct_pad(inputs, kernel_size):
+    """Returns a tuple for zero-padding for 2D convolution with downsampling.
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:307]: Space between every section here and in the other method docstrings

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+    kernel_size=3,
+    strides=1,
+    activation="swish",
+    expand_ratio=1,
+    se_ratio=0.0,
+    id_skip=True,
+    drop_rate=0.0,
+    name="",
+):
+    """An inverted residual block.
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:374]: This is a correctly spaced docstring 👍

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+    Args:
+        inputs: Tensor, The input tensor of the block
+        filters_in: integer, the number of input filters.
+        filters_out: integer, the number of output filters.
+        kernel_size: integer, the dimension of the convolution window.
+        strides: integer, the stride of the convolution.
+        activation: activation function.
+        expand_ratio: integer, scaling coefficient for the input filters.
+        se_ratio: float between 0 and 1, fraction to squeeze the input filters.
+        id_skip: boolean.
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:385]: Update as above

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        drop_rate: float between 0 and 1, fraction of the input units to drop.
+        name: string, block label.
+
+    Returns:
+        tf.Tensor
+    """
+    filters = filters_in * expand_ratio
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:392]: Set `x = inputs` up here and skip the `else`

**Review by jbischof — CHANGES_REQUESTED** [review]
Thank you, mostly small changes!

**Comment by jbischof** [comment]
@cosmo3769 can you do me a favor and run

```sh
pytest keras_cv/models/backbones/efficientnet_v1 --run_extra_large
```

to check all the presets locally?

You'll also need a merge commit before we can merge.

**Comment by cosmo3769** [comment]
Okay, working on these changes.

**Comment by cosmo3769** [comment]
> @cosmo3769 can you do me a favor and run
> 
> ```shell
> pytest keras_cv/models/backbones/efficientnet_v1 --run_extra_large
> ```
> 
> to check all the presets locally?
> 
> You'll also need a merge commit before we can merge.

Sure.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py (hunk 5 above)
Comment (cosmo3769) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:162]: True. Both gives same result while one giving in shorter code.

**Comment by cosmo3769** [comment]
> @cosmo3769 can you do me a favor and run
> 
> ```shell
> pytest keras_cv/models/backbones/efficientnet_v1 --run_extra_large
> ```
> 
> to check all the presets locally?

@jbischof Here is the output of the test result:
 
Screenshot 2023-06-11 at 8 41 37 PM

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py (hunk 4 above)
Comment (cosmo3769) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:45]: Should I write something like? 

width_coefficient: float, controls the scaling coefficient for the network width. Increasing this value will result in a wider network, while decreasing it will result in a narrower network.

depth_coefficient: float, controls the scaling coefficient for the network depth. Increasing this value will result in a deeper network, while decreasing it will result in a shallower network.

dropout_rate: float, specifies the dropout rate before the final classifier layer. It helps prevent overfitting.

drop_connect_rate: float, specifies the dropout rate at skip connections. Setting a dropout rate for skip connections helps regularize the network and prevent overfitting.

depth_divisor: integer, determines a unit of network width. It is used to ensure that the number of filters in each layer is divisible by this value and is used for achieving better computational efficiency.

**Comment by cosmo3769** [comment]
> Thanks @cosmo3769! Can you mimic the changes in #1873 into this and your other recent backbone PRs? Our old tests weren't catching some bugs.

Sure.

**Comment by cosmo3769** [comment]
Hi @jbischof, 

Here is the updated test output: 

Screenshot 2023-06-14 at 7 02 12 PM


[Colab link](https://colab.research.google.com/drive/16JeflSq2fBkwkmVbg3Mam9F3FjdY0Odr?usp=sharing) for layer matching. Some of the names are changed as to have uniqueness in name. The colab code is referenced from [here](https://colab.research.google.com/drive/1hzxwTgUgyiGp3YRZbMjB1FBcSuVXy-LY?usp=sharing).

**Comment by innat** [comment]
Any update on this PR? 
This bacakbone is required to progress this ticket https://github.com/keras-team/keras-cv/issues/951

**Comment by cosmo3769** [comment]
> Any update on this PR?
> This bacakbone is required to progress this ticket #951

The changes required for this backbone is done as per the reviews. Waiting for the final review on this. If there will be more changes required, I will do so and it will be ready to be merged. @innat

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_aliases.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        kwargs.update(
+            {
+                "include_rescaling": include_rescaling,
+                "input_shape": input_shape,
+                "input_tensor": input_tensor,
+            }
+        )
+        return EfficientNetV1Backbone.from_preset("efficientnetv1_b1", **kwargs)
+
+    @classproperty
```
[the remark this answers is not in the document]
Comment (jbischof) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_aliases.py:92]: Sorry if `presets_with_weights` is empty, so should `presets` be empty! You can check the other backbones for style guidance.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py (hunk 4 above)
Comment (jbischof) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:45]: Hey, pretty good! I was actually only talking about `depth_divisor`. Your clarification is helpful, but I still don't know if a user reading this would know good values to use.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+        include_rescaling,
+        width_coefficient,
+        depth_coefficient,
+        dropout_rate=0.2,
+        drop_connect_rate=0.2,
+        depth_divisor=8,
+        activation="swish",
+        input_shape=(None, None, 3),
+        input_tensor=None,
+        stackwise_kernel_sizes,
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:112]: don't args with defaults always need to come after args without them?

Regardless, let's put these stackwise args first to match other backbones.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+def apply_efficientnet_block(
+    inputs,
+    filters_in=32,
+    filters_out=16,
+    kernel_size=3,
+    strides=1,
+    activation="swish",
+    expand_ratio=1,
+    se_ratio=0.0,
+    drop_rate=0.0,
```
Comment (jbischof) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:364]: We've standardized on `dropout_rate` across the repo.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py (hunk 4 above)
Comment (jbischof) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:45]: New idea: since this param is `8` for all presets maybe that should be the default value?

**Review by jbischof — COMMENTED** [review]
Looking good @cosmo3769 just some small changes needed now. Can you also do a merge commit to pick up some changes in our testing infra?

Sorry for the delay, we're in a big sprint right now 🪄

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py (hunk 4 above)
Comment (cosmo3769) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py]: Yeah, I think this will be good. Also, the default value for drop_connect_rate is 0.2. Adding this as well.

**Comment by cosmo3769** [comment]
> Can you also do a merge commit to pick up some changes in our testing infra?

Merge commit done. Please check.

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+
+    Args:
+        include_rescaling: bool, whether to rescale the inputs. If set to
+            True, inputs will be passed through a `Rescaling(1/255.0)` layer.
+        width_coefficient: float, scaling coefficient for network width.
+        depth_coefficient: float, scaling coefficient for network depth.
+        dropout_rate: float, dropout rate before final classifier layer.
+        drop_connect_rate: float, dropout rate at skip connections. The default
+            value is set to 0.2.
+        depth_divisor: integer, a unit of network width. The default value is
```
Comment (cosmo3769) [inline keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_backbone.py:46]: I added like this. Did I understand you correctly by your [comment](https://github.com/keras-team/keras-cv/pull/1716#discussion_r1248066253)? Please correct me if I am wrong. Thank you.

**Comment by cosmo3769** [comment]
Porting is done as discussed here #1941. Please check. Thank you.

**Review by jbischof — CHANGES_REQUESTED** [review]
Thanks @cosmo3769 just one nit about the alias classes

File: keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_aliases.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (cosmo3769) [reply keras_cv/models/backbones/efficientnet_v1/efficientnet_v1_aliases.py:88]: Yeah, sorry. I missed to push the code changes I did in my local. Now I updated.

**Review by jbischof — APPROVED** [review]
Thanks for your patience and dedication on this PR!

**Comment by cosmo3769** [comment]
@jbischof Should I resolve the conflicts?
